[FX] Add delegate interface

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42865 [FX] Add delegate interface**

This adds an argument, `delegate` to `symbolic_trace` that allows the user to specify a delegate object to override certain behavior during symbolic tracing. Currently, the two behaviors that can be overridden are 1) leaf module identification and 2) the behavior of inserting a node. The latter can be used, for example, to exclude in-place operations from the trace.

Overriding `create_node` allows one to throw an exception when an erroneous condition is encountered. This exception will show a stack trace through the user's code indicating where the erroneous operation occured:

```
Traceback (most recent call last):
  File "test/test_fx.py", line 109, in test_disallow_override
    symbolic_trace(m2, delegate=NoMutableCallDelegate())
  File "/data/users/jamesreed/pytorch/torch/fx/symbolic_trace.py", line 108, in symbolic_trace
    graph.output(fn(*args))
  File "test/test_fx.py", line 105, in forward
    torch.log_(x)
  File "/data/users/jamesreed/pytorch/torch/fx/node.py", line 57, in __torch_function__
    self.graph, 'call_function', orig_method, args, kwargs, name=self.graph._name(orig_method.__name__))
  File "test/test_fx.py", line 88, in create_node
    raise RuntimeError('In-place operations are not supported')
RuntimeError: In-place operations are not supported

```

Differential Revision: [D23056584](https://our.internmc.facebook.com/intern/diff/D23056584)

Commit:
[FX] Add interface to reject nodes

[ghstack-poisoned]

Commit:
Update on "[FX] Add interface to reject nodes"




[ghstack-poisoned]

Commit:
Update on "[FX] Add interface to reject nodes"


Differential Revision: [D23056584](https://our.internmc.facebook.com/intern/diff/D23056584)

[ghstack-poisoned]

Commit:
Update on "[FX] Add interface to reject nodes"


This adds an argument, `is_valid_call`, to `symbolic_trace` that allows the caller to specify a function to gate certain callsites based on their `target`, `args`, and `kwargs`. This can be used in scenarios where a pass has a contract with the user that they don't support, for example, mutable operations.

Upon finding that a callsite is invalid, the caller-defined function should throw an exception describing why this call is rejected. Normal Python exception propagation will show a stack trace through the user code, for example:

```
Traceback (most recent call last):
  File "test/test_fx.py", line 111, in test_disallow_override
    symbolic_trace(m3, is_valid_call=no_mutable_is_valid_call)
  File "/data/users/jamesreed/pytorch/torch/fx/symbolic_trace.py", line 109, in symbolic_trace
    graph.output(fn(*args))
  File "test/test_fx.py", line 107, in forward
    y.add_(x)
  File "/data/users/jamesreed/pytorch/torch/fx/node.py", line 55, in __torch_function__
    self.is_valid_call(orig_method.__name__, args, kwargs)
  File "test/test_fx.py", line 81, in no_mutable_is_valid_call
    raise RuntimeError('In-place operations are not supported')
RuntimeError: In-place operations are not supported

```

Differential Revision: [D23056584](https://our.internmc.facebook.com/intern/diff/D23056584)

[ghstack-poisoned]

rdzhabarov commented on torch/fx/symbolic_trace.py:
```diff
@@ -43,15 +44,40 @@ def _patch_function(fn, nargs):
+def symbolic_trace(root : torch.nn.Module,
+                   is_leaf_module : Callable[[torch.nn.Module], bool] = is_leaf_module,
+                   is_valid_call : Callable[
+                       [Union[str, Callable], Tuple[Any]], Dict[str, Any]] = default_is_valid_call):
```
we'd probably need to provide some helpers for common scenarios, no in-place, no dict args, etc.

does not need to be in this PR.

rdzhabarov reviewed (approved):
👍  nice!

Commit:
Update on "[FX] Add interface to reject nodes"


This adds an argument, `is_valid_call`, to `symbolic_trace` that allows the caller to specify a function to gate certain callsites based on their `target`, `args`, and `kwargs`. This can be used in scenarios where a pass has a contract with the user that they don't support, for example, mutable operations.

Upon finding that a callsite is invalid, the caller-defined function should throw an exception describing why this call is rejected. Normal Python exception propagation will show a stack trace through the user code, for example:

```
Traceback (most recent call last):
  File "test/test_fx.py", line 111, in test_disallow_override
    symbolic_trace(m3, is_valid_call=no_mutable_is_valid_call)
  File "/data/users/jamesreed/pytorch/torch/fx/symbolic_trace.py", line 109, in symbolic_trace
    graph.output(fn(*args))
  File "test/test_fx.py", line 107, in forward
    y.add_(x)
  File "/data/users/jamesreed/pytorch/torch/fx/node.py", line 55, in __torch_function__
    self.is_valid_call(orig_method.__name__, args, kwargs)
  File "test/test_fx.py", line 81, in no_mutable_is_valid_call
    raise RuntimeError('In-place operations are not supported')
RuntimeError: In-place operations are not supported

```

Differential Revision: [D23056584](https://our.internmc.facebook.com/intern/diff/D23056584)

[ghstack-poisoned]

Commit:
Update on "[FX] Add interface to reject nodes"


This adds an argument, `is_valid_call`, to `symbolic_trace` that allows the caller to specify a function to gate certain callsites based on their `target`, `args`, and `kwargs`. This can be used in scenarios where a pass has a contract with the user that they don't support, for example, mutable operations.

Upon finding that a callsite is invalid, the caller-defined function should throw an exception describing why this call is rejected. Normal Python exception propagation will show a stack trace through the user code, for example:

```
Traceback (most recent call last):
  File "test/test_fx.py", line 111, in test_disallow_override
    symbolic_trace(m3, is_valid_call=no_mutable_is_valid_call)
  File "/data/users/jamesreed/pytorch/torch/fx/symbolic_trace.py", line 109, in symbolic_trace
    graph.output(fn(*args))
  File "test/test_fx.py", line 107, in forward
    y.add_(x)
  File "/data/users/jamesreed/pytorch/torch/fx/node.py", line 55, in __torch_function__
    self.is_valid_call(orig_method.__name__, args, kwargs)
  File "test/test_fx.py", line 81, in no_mutable_is_valid_call
    raise RuntimeError('In-place operations are not supported')
RuntimeError: In-place operations are not supported

```

Differential Revision: [D23056584](https://our.internmc.facebook.com/intern/diff/D23056584)

[ghstack-poisoned]

zdevito reviewed (approved):
Looks good. This suggests a follow-up refactor that separates Node from the tracing process that I will submit as a follow-up PR

Commit:
Update on "[FX] Add delegate interface"


This adds an argument, `is_valid_call`, to `symbolic_trace` that allows the caller to specify a function to gate certain callsites based on their `target`, `args`, and `kwargs`. This can be used in scenarios where a pass has a contract with the user that they don't support, for example, mutable operations.

Upon finding that a callsite is invalid, the caller-defined function should throw an exception describing why this call is rejected. Normal Python exception propagation will show a stack trace through the user code, for example:

```
Traceback (most recent call last):
  File "test/test_fx.py", line 111, in test_disallow_override
    symbolic_trace(m3, is_valid_call=no_mutable_is_valid_call)
  File "/data/users/jamesreed/pytorch/torch/fx/symbolic_trace.py", line 109, in symbolic_trace
    graph.output(fn(*args))
  File "test/test_fx.py", line 107, in forward
    y.add_(x)
  File "/data/users/jamesreed/pytorch/torch/fx/node.py", line 55, in __torch_function__
    self.is_valid_call(orig_method.__name__, args, kwargs)
  File "test/test_fx.py", line 81, in no_mutable_is_valid_call
    raise RuntimeError('In-place operations are not supported')
RuntimeError: In-place operations are not supported

```

Differential Revision: [D23056584](https://our.internmc.facebook.com/intern/diff/D23056584)

[ghstack-poisoned]

dzhulgakov commented on test/test_fx.py:
```diff
@@ -73,5 +78,65 @@ def forward(self, d):
+                name = target if isinstance(target, str) else torch.typename(target)
+                if name[-1] == '_':
+                    raise RuntimeError('In-place operations are not supported')
+                return super().create_node(kind, target, args, kwargs, name)
```
nit: it seems to miss `graph` argument, but still works because we throw an error first

dzhulgakov commented on torch/fx/node.py:
```diff
@@ -20,14 +20,15 @@ def __init__(self, graph, name, op, target, args, kwargs):
+        return Attribute(self, k, self.delegate)
 
     def __call__(self, *args, **kwargs):
         return self._create_node('call_method', '__call__', [self] + args, kwargs)
```
should this one also be using delegate.create_node?
